app/service/data_svc.py

```diff
@@ -134,41 +134,17 @@ def __init__(self):
             self.log.error('[!] REMOVE: %s' % e)
 
     """ PRIVATE """
-
-    @staticmethod
-    async def _add_phase_abilities(phase_dict, phase, phase_entries):
-        for ability in phase_entries:
-            phase_dict[phase].append(ability)
-        return phase_dict
-
-    async def _insert_pack_phases(self, pack, phases, current_phase, adversary):
-        phases_new = await self._add_adversary_packs(pack)
-        if phases_new:
-            for i, phase in phases_new.items():
-                phases.insert(current_phase + i, phase)
-            return current_phase + i
-        else:
-            self.log.error(
-                'Missing ability or pack (%s) for adversary: %s (%s)' % (pack, adversary['name'], adversary['id']))
-            return 0
-
-    async def _add_phases(self, phases, adversary):
-        pp = defaultdict(list)
-        phase_id = 0
-        while phase_id < len(phases):
-            for idx, step in enumerate(phases[phase_id]):
-                abilities = await self.locate('abilities', match=dict(ability_id=step))
-                if abilities:
-                    await self._add_phase_abilities(pp, phase_id + 1, abilities)
-                else:
-                    # insert this phase and shift down later abilities to new phase
-                    del phases[phase_id][idx]
-                    last_phase = await self._insert_pack_phases(step, phases, phase_id, adversary)
-                    if last_phase and idx < len(phases[phase_id]):
-                        phases.insert(last_phase + 1, [phases[phase_id][idx]])
-                        del phases[phase_id][idx + 1:]
-            phase_id += 1
-        return dict(pp)
+    async def _link_abilities(self, ordering, adversary):
+        atomics = []
+        for ability in ordering:
+            trigger = await self.locate('abilities', match=dict(ability_id=ability))
+            if trigger:
+                for variant in trigger:
+                    atomics.append(variant)
+            else:
+                self.log.error(
+                    'Missing ability (%s) for adversary: %s (%s)' % (ability, adversary['name'], adversary['id']))
```

Comment:
should this be a list comprehension and try/except block?

Reply:
We could write a really long list comprehension to do this in one go, but I've had a hard time with using async in list comprehensions, so I'm inclined to leave it as is (which is similar to the original design). If you'd prefer me trying to rework this logic, I can do so, though.

Reply:
try:

```python
async def _link_abilities(self, ordering, adversary):
    try:
        return [v for ability in ordering for v in await self.locate('abilities', match=dict(ability_id=ability))]
    except Exception as e:
        self.log.error('Could not build ability list: %s' % e)
```